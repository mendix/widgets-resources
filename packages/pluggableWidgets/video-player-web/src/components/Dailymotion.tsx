import { Component, createElement, createRef } from "react";
import ReactResizeDetector from "react-resize-detector";
import { fixHeightWithRatio, getRatio, validateUrl } from "../utils/Utils";

export interface DailymotionProps {
    url: string;
    autoPlay: boolean;
    showControls: boolean;
    muted: boolean;
    aspectRatio?: boolean;
}

export interface DailymotionState {
    ratio: number;
}

export class Dailymotion extends Component<DailymotionProps, DailymotionState> {
    private iframe = createRef<HTMLIFrameElement>();
    private readonly handleOnResize = this.onResize.bind(this);
    private handleAttributes = this.getUrlAttributes.bind(this);
    readonly state: DailymotionState = {
        ratio: 0
    };

    render(): JSX.Element {
        return (
            <iframe
                className="widget-video-player-iframe"
                src={this.generateUrl(this.props.url)}
                frameBorder="0"
                allow="autoplay; fullscreen"
                allowFullScreen
                ref={this.iframe}
            >
                <ReactResizeDetector
                    handleWidth
                    handleHeight
                    onResize={this.handleOnResize}
                    refreshMode="debounce"
                    refreshRate={100}
                />
            </iframe>
        );
    }

    private onResize(): void {
        if (this.iframe && this.iframe.current && this.props.aspectRatio) {
            if (this.state.ratio) {
                fixHeightWithRatio(this.iframe.current, this.state.ratio);
            } else {
                getRatio(this.props.url).then(ratio => {
                    this.setState({ ratio });
                    if (this.iframe && this.iframe.current) {
                        fixHeightWithRatio(this.iframe.current, this.state.ratio);
                    }
                });
            }
        }
    }

    private generateUrl(url: string): string {
        const attributes = this.handleAttributes();
        try {
            if (url.includes("dailymotion.com/embed")) {
                return `${url}${attributes}`;
            }

            const urlVimeoSplit = url.split("/");
            if (urlVimeoSplit.length > 0) {
                const id = urlVimeoSplit[urlVimeoSplit.length - 1];
                if (id) {
                    return `https://www.dailymotion.com/embed/video/${id}${attributes}`;
                }
            }
        } catch (e) {
            return url;
        }
        return url;
    }

    private getUrlAttributes(): string {
        let attributes = "?sharing-enable=false";
        attributes += "&autoplay=" + (this.props.autoPlay ? "true" : "false");
        attributes += "&mute=" + (this.props.muted ? "true" : "false");
        attributes += "&controls=" + (this.props.showControls ? "true" : "false");
        return attributes;
    }

    // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
    public static canPlay(url: string): boolean {
        return !!url && !!validateUrl(url) && url.indexOf("dailymotion.com") > -1;
    }
}
