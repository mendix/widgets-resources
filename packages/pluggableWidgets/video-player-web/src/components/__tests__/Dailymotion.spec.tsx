import { createElement } from "react";

import { Dailymotion, DailymotionProps } from "../Dailymotion";
import { create } from "react-test-renderer";
import ReactResizeDetector from "react-resize-detector";

describe("Dailymotion Player", () => {
    const defaultProps = {
        url: "test",
        autoPlay: false,
        muted: false,
        showControls: false,
        aspectRatio: false
    };

    const defaulPlayer = (props: DailymotionProps): JSX.Element => <Dailymotion {...props} />;

    it("should renders correctly", () => {
        const player = create(defaulPlayer(defaultProps)).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with autoplay", () => {
        const player = create(defaulPlayer({ ...defaultProps, autoPlay: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with muted", () => {
        const player = create(defaulPlayer({ ...defaultProps, muted: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with controls", () => {
        const player = create(defaulPlayer({ ...defaultProps, showControls: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with aspectRatio", () => {
        const player = create(defaulPlayer({ ...defaultProps, aspectRatio: true }));
        window.dispatchEvent(new Event("resize"));
        const instance = player.root;
        const sizeDetector = instance.findByType(ReactResizeDetector);

        expect(player.toJSON).toMatchSnapshot();
        expect(sizeDetector).not.toBeNull();
        expect(sizeDetector.props).toHaveProperty("onResize");
    });
});
