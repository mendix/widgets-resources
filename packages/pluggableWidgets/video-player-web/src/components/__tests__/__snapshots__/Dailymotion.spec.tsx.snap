// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dailymotion Player should renders correctly 1`] = `
<iframe
  allow="autoplay; fullscreen"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://www.dailymotion.com/embed/video/test?sharing-enable=false&autoplay=false&mute=false&controls=false"
>
  <div />
</iframe>
`;

exports[`Dailymotion Player should renders correctly with aspectRatio 1`] = `[Function]`;

exports[`Dailymotion Player should renders correctly with autoplay 1`] = `
<iframe
  allow="autoplay; fullscreen"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://www.dailymotion.com/embed/video/test?sharing-enable=false&autoplay=true&mute=false&controls=false"
>
  <div />
</iframe>
`;

exports[`Dailymotion Player should renders correctly with controls 1`] = `
<iframe
  allow="autoplay; fullscreen"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://www.dailymotion.com/embed/video/test?sharing-enable=false&autoplay=false&mute=false&controls=true"
>
  <div />
</iframe>
`;

exports[`Dailymotion Player should renders correctly with muted 1`] = `
<iframe
  allow="autoplay; fullscreen"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://www.dailymotion.com/embed/video/test?sharing-enable=false&autoplay=false&mute=true&controls=false"
>
  <div />
</iframe>
`;
